function_calling/src/tools.py

```diff
@@ -0,0 +1,163 @@
+import os
+import re
+import sys
+import operator
+from datetime import datetime
+from dotenv import load_dotenv
+from typing import  Optional, Union
+from langchain_core.tools import StructuredTool, ToolException, Tool,tool
+from langchain_community.tools.sql_database.tool import QuerySQLDataBaseTool
+from langchain_core.pydantic_v1 import BaseModel, Field
+from langchain_experimental.utilities import PythonREPL
+from langchain_community.llms.sambanova import SambaStudio
+from langchain_community.utilities import SQLDatabase
+from langchain.chains import create_sql_query_chain
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
+kit_dir = os.path.abspath(os.path.join(current_dir, ".."))
+repo_dir = os.path.abspath(os.path.join(kit_dir, ".."))
+sys.path.append(kit_dir)
+sys.path.append(repo_dir)
+
+load_dotenv(os.path.join(repo_dir, ".env")) 
+
+##Get time tool
+
+# tool schema
+class GetTimeSchema(BaseModel):
+    """Returns current date, current time or both."""
+    kind: Optional[str] = Field(description='kind of information to retrieve "date", "time" or "both"')
+    
+# definition using @tool decorator
+@tool(args_schema=GetTimeSchema)
+def get_time(kind: str = "both") -> str:
+    """Returns current date, current time or both.
+
+    Args:
+        kind: date, time or both
+    """
+    if kind == "date":
+        date = datetime.now().strftime("%d/%m/%Y")
+        return f"Current date: {date}"
+    elif kind == "time":
+        time = datetime.now().strftime("%H:%M:%S")
+        return f"Current time: {time}"
+    else:
+        date = datetime.now().strftime("%d/%m/%Y")
+        time = datetime.now().strftime("%H:%M:%S")
+        return f"Current date: {date}, Current time: {time}"
+
+## Calculator Tool
+
+class CalculatorSchema(BaseModel):
+    """allow calculation of only basic operations: + - * and /
+    with a string input expression """
+    expression: str = Field(..., description="expression to calculate, example '12 * 3'")
+    
+#function to use in the tool
+def calculator(expression: str ) -> Union[str, int, float]:
+    """
+    allow calculation of basic operations
+    with a string input expression
+    Args:
+        expression: expression to calculate 
+    """
+    ops = {
+        '+': operator.add,
+        '-': operator.sub,
+        '*': operator.mul,
+        'x': operator.mul,
+        'X': operator.mul,
+        '÷': operator.truediv,
+        '/': operator.truediv
+    }
+    tokens = re.findall(r'\d+\.?\d*|\+|\-|\*|\/|÷|x|X', expression)
+    
+    if not tokens:
+        raise ToolException(f"Invalid expression '{expression}', should only contain one of the following operators + - * and /")
```

Comment:
I would add all the remaining operators as well to the message that are in ops.

Reply:
added